### PR TITLE
Defer selection of triton template

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120278
* #121998
* #121997
* __->__ #120275
* #121996

Our prior approach to epilogue fusion was to select from a choice from a set of triton templates and extern calls based on benchmarking inputs, then unconditionally fuse epilogues. This can be sub-optimal in following ways:

- We select an extern kernel, however an epilogue like relu() exists such that choosing a triton template + relu would have been faster
- We select a triton template, epilogue fuse, and register spilling occurs causing it to be slower than not epilogue fusing.

In this PR we wait to select either the Triton Template or Extern Kernel based on benchmarking results from the kernel itself and its epilogue. As soon as a successful fusion occurs where a fused Triton Template + epilogue is faster than the unfused choice we finalize the MultiTemplateBuffer as a specific template. If no fusion occurs we'll finalize the MultiTemplateBuffer after fusion. 

Note: if there are multiple epilogue fusions (not super likely), even though we select a template after the first fusion, we will still benchmark to see if subsequent epilogue are worth fusing. We could potentially defer choosing template in this case in a follow up at expense of compile time.

Gives 4% HF training win, 10% TIMM inference win. Increases compilation time which I will be trying to address more in follow up prs.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by jansel — APPROVED** [review]
@shunting314 please take a look at this too

**Comment by shunting314** [comment]
> Gives 4% HF training win, 10% TIMM inference win. Increases compilation time which I will be trying to address more in follow up prs.

Nice!
A couple of questions before I look more into the code:
1. is this comparing with enable MaxAutotune but without the PR?
2. I'm a bit curious how much win we can get by the previous PR only ( https://github.com/pytorch/pytorch/pull/118880 )

**Comment by eellison** [comment]
@shunting314 yes this is comparing with max autotune. I haven't benchmark'd just the benchmark templates yet. I suspect most of the win is in choosing triton templates we wouldnt have have instead of not fusing epilogues (so mostly this pr).

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -3839,6 +3851,12 @@ def store_cache():
             # generating out of range indices for later calls.
             ms = do_bench(lambda: call(wrapped_jit_function.clone_args(*args)[0]))
 
+            # overhead of cloning args gives bias for fusing the kernel
+            # in the case of mutating/in-placeable second fusion
+            # TODO - would be better as a hook in triton do_bench that reset
+            # the input values between benchmarking
+            ms = ms - do_bench(lambda: wrapped_jit_function.clone_args(*args))
```
Comment (shunting314) [inline torch/_inductor/codegen/triton.py:3858]: The cloning overhead should be the same for different choices anyway? That should not affect comparison of kernel latencies

**Comment by shunting314** [comment]
> yes this is comparing with max autotune.

I see. I saw benchmark fusion is enabled as well. So some of the gain should be due to benchmark fusion?

File: torch/_inductor/select_algorithm.py [hunk 2]
```diff
@@ -842,6 +846,40 @@ def autotune(choices):
...
+                    min_extern_choice = min(min_extern_choice, timing)
+
+            timings = {
+                choice: time
+                for choice, time in timings.items()
+                if (
+                    time <= min_extern_choice
+                    or not isinstance(choice, ExternKernelCaller)
+                )
+            }
```
Comment (shunting314) [inline torch/_inductor/select_algorithm.py:863]: hmm, this is no-op? Since we only have one ExternKernelCaller

File: torch/_inductor/scheduler.py [hunk 3]
```diff
@@ -1722,42 +1770,82 @@ def speedup_by_fusion(self, node1, node2):
...
+                        node1.get_names(),
+                        node2.get_names(),
+                        red_text(f"{ms_fused / (ms1 + ms2):.3f}"),
+                    )
+
+        if isinstance(node1, SchedulerNode) and isinstance(
+            node1.node, ir.MultiTemplateBuffer
+        ):
+            multi_node = node1.node
+            choice_timings = multi_node.choice_timings
```
Comment (shunting314) [inline torch/_inductor/scheduler.py:1794]: hmm, do you need update multi_node.choice_timings with the latency including the fused epilogue?
Let's say we have a sequence: matmul, pointwise1, pointwise2.

Currently code seems to use the latency of matmul rather than (matmul + pointwise1) when we check if  further fusing in  pointwise2 is beneficial

File: torch/_inductor/select_algorithm.py (hunk 2 above)
Comment (eellison) [reply torch/_inductor/select_algorithm.py:863]: in `tuned_addmm` we select from multiple options

File: torch/_inductor/codegen/triton.py (hunk 1 above)
Comment (shunting314) [reply torch/_inductor/codegen/triton.py:3858]: But most likely those in-place args should also become inplace args for the fused 'template + epilogue'?

It's fine to add this to be more safe though

Comment (eellison) [reply torch/_inductor/codegen/triton.py:3858]: In the case of mm + relu the inplacing is on the output of template

**Comment by eellison** [comment]
> > yes this is comparing with max autotune.
> 
> 
> 
> I see. I saw benchmark fusion is enabled as well. So some of the gain should be due to benchmark fusion?

I'll run a benchmark to be sure. I only enabled benchmark fusion on templates in the run, not otherwise

File: torch/_inductor/codegen/triton.py (hunk 1 above)
Comment (shunting314) [reply torch/_inductor/codegen/triton.py:3858]: ohh, that's a good point

File: torch/_inductor/scheduler.py (hunk 3 above)
Comment (eellison) [reply torch/_inductor/scheduler.py:1794]: We actually only go through this process once, and then we finalize the template. I'll add comments / make clearer
